libbpf-tools/readahead.bpf.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "readahead.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, u8);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} in_readahead SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct page *);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} birth SEC(".maps");
+
+static struct hist hist;
+
+SEC("fentry/__do_page_cache_readahead")
+int BPF_PROG(fentry____do_page_cache_readahead)
```

Comment:
btw, you don't have to folow this convention of duplicating program type in function name, it's actually even better to have shorter function name, so maybe try just `do_page_cache_readahead`. bpftool will show that it's an fentry program anyways.

Reply:
Got it, thanks!